### PR TITLE
Rajkumarpatel2602 kv benchmarking

### PR DESCRIPTION
**Pull request by rajkumarpatel2602**
This commit includes Benchmarking test and json to csv parser for daos KVS for various workloads.

kv_benchmarking.cc contains Put, Get, List and Remove functionalities for various workloads (1K, 10K, 100K, 1M, etc. Key-value operations for each functionalities with various key-size and value-size combinations).

json_report_to_csv.py includes actions for generating favorable csv file using the json file which is generated at the end of benchmarking test.

### CODE REVIEW

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#define POOL_ID "7d630b52-65c3-4061-aa87-4e02b4e6d818"
+
+/* number of keys */
```
Comment (kaustubh-d) [inline daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:17]: Number of ops (operations)

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+#define NR_KEYS_100    100
+#define NR_KEYS_1000   1000
+#define NR_KEYS_10000  10000
+#define NR_KEYS_100000 100000
+
+/* number of query at a time while listing */
+#define NR_QUERY 10
```
Comment (kaustubh-d) [inline daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:24]: NR_KV_PER_LISTING

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+
+/* value sizes */
+#define BM_VAL_1K  1024
+#define BM_VAL_4K  (1024 * 4)
+#define BM_VAL_8K  (1024 * 8)
+#define BM_VAL_16K (1024 * 16)
+#define BM_VAL_32K (1024 * 32)
+
+/* buffer to hold keys while listing */
+#define KEY_DESC_BUF (1024 * 1024)
```
Comment (kaustubh-d) [inline daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:41]: KEY_LIST_BUF

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+
+#define ARG_KEY_SIZE_OPTIONS            { BM_KEY_64B , BM_KEY_128B, BM_KEY_256B, BM_KEY_512B, BM_KEY_1024B }
+#define ARG_VAL_SIZE_OPTIONS            { BM_VAL_1K , BM_VAL_4K, BM_VAL_8K, BM_VAL_16K, BM_VAL_32K }
+#define NR_KEYS_PER_KV_SIZE_COMBINATION { NR_KEYS_100 /*,NR_KEYS_1000, NR_KEYS_10000, NR_KEYS_100000 */ }
```
Comment (kaustubh-d) [inline daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:45]: NR_OPS_OPTIONS

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+
+#define ARG_MATRICS\
+   { ARG_KEY_SIZE_OPTIONS, ARG_VAL_SIZE_OPTIONS, NR_KEYS_PER_KV_SIZE_COMBINATION }
+
+static char node[ 128 ] = "new_node";
```
Comment (kaustubh-d) [inline daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:50]: err_msg, remove value.

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+
+static daos_handle_t poh;
```
Comment (kaustubh-d) [inline daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:52]: add comment, pool handle

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+        exit(1);                                                    \
+    } while (0)
+
+#define ASSERT( cond, ... )                                         \
+    do {                                                            \
+        if (!(cond))                                                \
+        FAIL(__VA_ARGS__);                                          \
+    } while (0)
+
+#define BUFLEN 100
```
Comment (kaustubh-d) [inline daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:68]: remove

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+
+daos_handle_t oh;
+daos_obj_id_t oid;
+int           rc;
```
Comment (kaustubh-d) [inline daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:72]: we dont need global rc

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+   rc = daos_cont_create( poh, co_uuid, NULL /* properties */,
+                          NULL /* event */ );
+   ASSERT( rc == 0, "container create failed with %d", rc );
+
+   /** open container */
+   rc = daos_cont_open( poh, co_uuid, DAOS_COO_RW, &coh, NULL,
+                        NULL );
+   ASSERT( rc == 0, "container open failed with %d", rc );
+
+   /** share container handle with peer tasks */
```
Comment (kaustubh-d) [inline daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:101]: remove this comment

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+
+   rc = daos_pool_disconnect( poh, NULL );
+   ASSERT( rc == 0, "disconnect failed" );
+
+   /** teardown the DAOS stack */
+   rc = daos_fini( );
+   ASSERT( rc == 0, "daos_fini failed with %d", rc );
+}
+
+void gen_key_names( char *key_buf, int nr, int key_size )
```
Comment (kaustubh-d) [inline daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:128]: Add comment about key_name format generated by this method.

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc [hunk 11]
```diff
@@ -0,0 +1,400 @@
...
+      0
+   };
+
+   /* generate different key */
+   memset( key_buf, 'x', key_size - 1 );
+   sprintf( key_name, "%.16d", nr );
+   strncpy( ( char * )key_buf + strlen( key_buf ) - strlen( key_name ), ( char * )key_name, strlen( key_name ) );
+}
+
+// bnechmark function is getting called here.
```
Comment (kaustubh-d) [inline daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:140]: s/bnechmark/benchmark

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc [hunk 12]
```diff
@@ -0,0 +1,400 @@
...
+static void kv_remove_function( benchmark::State &state ) {
+   /* perform setup */
+   setup_main( );
+
+   unsigned int key_size = state.range( 0 );
+   unsigned int val_size = state.range( 1 );
+   unsigned int num_keys = state.range( 2 );
```
Comment (kaustubh-d) [inline daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:147]: nr_ops

Comment (kaustubh-d) [reply daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:147]: comment, nr_ops == num of key-value pairs for test

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc [hunk 13]
```diff
@@ -0,0 +1,400 @@
...
+      {
+         state.PauseTiming( );
+
+         /* generate different key */
+         gen_key_names( key_buf, i, key_size );
+
+         state.ResumeTiming( );
+
+         /* actual function to mearsure time */
+         daos_kv_put( oh, DAOS_TX_NONE, 0, ( char * )key_buf, key_size, val_buf, NULL );
```
Comment (kaustubh-d) [inline daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:307]: replace key_size with size of value buffer

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc [hunk 14]
```diff
@@ -0,0 +1,400 @@
...
+      char            *buf;
+      daos_key_desc_t kds[ NR_QUERY ];
+      daos_anchor_t   anchor = {
+         0
+      };
+      int             key_nr = 0;
+      d_sg_list_t     sgl;
+      d_iov_t         sg_iov;
+
+      buf = ( char * )calloc( KEY_DESC_BUF, sizeof( char ) );
```
Comment (kaustubh-d) [inline daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:216]: list_of_keys

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc [hunk 15]
```diff
@@ -0,0 +1,400 @@
...
+      d_iov_set( &sg_iov, buf, KEY_DESC_BUF );
+      sgl.sg_nr     = 1;
+      sgl.sg_nr_out = 0;
+      sgl.sg_iovs   = &sg_iov;
+
+      while ( !daos_anchor_is_eof( &anchor ) )
+      {
+         uint32_t nr = NR_QUERY;
+         int      rc;
+
```
Comment (kaustubh-d) [inline daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:226]: exclude memset from timing

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc [hunk 16]
```diff
@@ -0,0 +1,400 @@
...
+            continue;
+         }
+         else // if returned descriptors are non zero then compute each key and query value.
+         {
+            unsigned int offset = 0;
+
+            /* compute each key and fetch value */
+            for ( int i = 0; i < nr; i++ )
+            {
+               memset( key_buf, 0, key_size );
```
Comment (kaustubh-d) [inline daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:245]: exclude line 245 - 256 from timings to more accurately capture daos time

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc (hunk 13 above)
Comment (rajkumarpatel2602) [reply daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc]: done

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc (hunk 9 above)
Comment (rajkumarpatel2602) [reply daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc]: done

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc (hunk 1 above)
Comment (rajkumarpatel2602) [reply daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc]: done

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc (hunk 2 above)
Comment (rajkumarpatel2602) [reply daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc]: done

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc (hunk 3 above)
Comment (rajkumarpatel2602) [reply daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc]: done

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc (hunk 4 above)
Comment (rajkumarpatel2602) [reply daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc]: done

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc (hunk 5 above)
Comment (rajkumarpatel2602) [reply daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc]: done

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc (hunk 6 above)
Comment (rajkumarpatel2602) [reply daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc]: done

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc (hunk 8 above)
Comment (rajkumarpatel2602) [reply daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc]: done

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc (hunk 11 above)
Comment (rajkumarpatel2602) [reply daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc]: done

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc (hunk 16 above)
Comment (rajkumarpatel2602) [reply daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:328]: done

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc (hunk 15 above)
Comment (rajkumarpatel2602) [reply daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:309]: done

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc (hunk 14 above)
Comment (rajkumarpatel2602) [reply daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc]: done

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc (hunk 12 above)
Comment (rajkumarpatel2602) [reply daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc]: done

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc (hunk 10 above)
Comment (rajkumarpatel2602) [reply daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc]: done

File: daos-cortx/src/daos_kv_benchmark/json_report_to_csv.py [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+    str1 = str(list_num[0])
+    str2 = str(list_num[1])
+    str3 = str(list_num[2])
+
+    # create new fields - keysize_valsize and nr_ops
+    element['keysize_valsize']= str1 + str2
+    element['nr_ops']= str3
+    
+    # create more fields using existing data points
+    element['realtime_per_op_ms']=float(element['realtime_ms'])/int(element['nr_ops'])
```
Comment (kaustubh-d) [inline daos-cortx/src/daos_kv_benchmark/json_report_to_csv.py:49]: Please fix indentation

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc [hunk 18]
```diff
@@ -0,0 +1,448 @@
...
+#define DEBUG_LOG 0
+
+#if DEBUG_LOG
+#define LOG_MSG printf
+#else
+#define LOG_MSG //
+#endif
+
+static daos_handle_t poh; /* daos pool handle */
+static daos_handle_t coh; /* daos contianer handle */
```
Comment (kaustubh-d) [inline daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:63]: s/contianer/container/

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc [hunk 19]
```diff
@@ -0,0 +1,448 @@
...
+   ASSERT( rc == 0, "disconnect failed" );
+
+   /* teardown the DAOS stack */
+   rc = daos_fini( );
+   ASSERT( rc == 0, "daos_fini failed with %d", rc );
+}
+
+/**
+ * @brief This function generates key name of key_size using number given number
+ *        Key format geneated by this function would be a string of characters
```
Comment (kaustubh-d) [inline daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:148]: s/geneated/generated/

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc [hunk 20]
```diff
@@ -0,0 +1,448 @@
...
+
+   /* allocate key and value buffers */
+   char *key_buf = ( char * )calloc( key_size, sizeof( char ) ); // key buffer allocated
+   char *val_buf = ( char * )calloc( val_size, sizeof( char ) ); // value buffer allocated
+   memset( val_buf, 'z', val_size - 1 );                         // populate with some random value.
+
+   /* actual computation starts here */
+   for ( auto _ : state )
+   {
+      /* call daos_kv_put for state.range(2) times */
```
Comment (kaustubh-d) [inline daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:188]: s/for state.range(2) times/for num_ops times

Comment (kaustubh-d) [reply daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:188]: please check same in other functions below

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc [hunk 21]
```diff
@@ -0,0 +1,448 @@
...
+               offset += kds[ i ].kd_key_len;
+
+               state.ResumeTiming( );
+
+               /* compute value for the key_buf */
+               daos_size_t size = 0;
+               size = val_size;
+               daos_kv_get( oh, DAOS_TX_NONE, 0, key_buf, &size, rbuf, NULL );
+
+               //printf( "rbuf is %s\n", rbuf );
```
Comment (kaustubh-d) [inline daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:361]: remove or use LOG_MSG

Comment (rajkumarpatel2602) [reply daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc]: done

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc (hunk 20 above)
Comment (rajkumarpatel2602) [reply daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc]: done

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc (hunk 19 above)
Comment (rajkumarpatel2602) [reply daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc]: done

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc (hunk 18 above)
Comment (rajkumarpatel2602) [reply daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc]: done

File: daos-cortx/src/daos_kv_benchmark/json_report_to_csv.py (hunk 17 above)
Comment (rajkumarpatel2602) [reply daos-cortx/src/daos_kv_benchmark/json_report_to_csv.py]: done

File: daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc [hunk 22]
```diff
@@ -0,0 +1,441 @@
...
+#include <fcntl.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#define POOL_ID "2bd513ad-66d1-4312-829b-d69d8637b455" /* pre-created pool uuid */
```
Comment (kaustubh-d) [inline daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc:18]: Lets try gflags to make it a param.

Comment (rajkumarpatel2602) [reply daos-cortx/src/daos_kv_benchmark/kv_benchmarking.cc]: adopted logic to get pool uuid from environment as using gflags was not possible with existing setup.

File: daos-cortx/src/daos_kv_benchmark/README.md [hunk 23]
```diff
@@ -0,0 +1,16 @@
+This directory contains 3 files.
```
Comment (kaustubh-d) [inline daos-cortx/src/daos_kv_benchmark/README.md:1]: please mention prerequisites to use google benchmark and link to how to build and install.

**Comment by kaustubh-d** [comment]
@rajkumarpatel2602 please fix the DCO and Codacy errors.

**Review by johnbent — APPROVED** [review]

File: daos-cortx/src/daos_kv_benchmark/README.md (hunk 23 above)
Comment (rajkumarpatel2602) [reply daos-cortx/src/daos_kv_benchmark/README.md]: done

**Review by kaustubh-d — APPROVED** [review]
+1

**Pull request merged**